### ISSUE TITLE
ConnectionClosed never raised under heavy load?

### ISSUE THREAD

**Issue by satchamo** [aaugustin/websockets#849]
I use your (fantastic) websockets library to handle a workflow like this:

1. Clients connect, and get added to a list
2. Each client just hangs around via `async for message in websocket`
3. When a client gets a message, it just echos the message back to the client
4. The whole thing is wrapped in a try/catch. In the finally clause, I remove the client from my connection list

I'm simplifying things. But not by much (the code is less than 400 lines).

At around 12,000 active connections, something starts going wrong. The ClosedConnection exception does not seemed to get raised in a timely manner or at all. I have a hunch it _would_ be raised at some point (if I waited long enough), but I have to kill the process before it has time to recover (because it starts consuming all my memory). There is a lot of connection churn in my clients, and I can easily see 1k connections open and close within the span of a minute. To be clear, I typically only have 10k-12k "real" active connections. But because the clients are never "officially" disconnected in Python-land, I have a connection list that grows to 200k+ before consuming all my available memory, and I `kill -9` the proc.

I added some code that polled the list of connections every 5 seconds to see if any connections were in the `closed` state, but had not yet raised a closed exception (by virtue of still being in my connection list). And indeed, that is always the case when this situation happens. The connections were definitely disconnected from the system (verified with the `ss` command).

I was hoping someone more experienced in this domain would be able to recommend some things to check on (either in my code, or websockets or even in Python). 

websockets 8.1
Python 3.6.9
Ubuntu 18.04 (4.15.0-96)
Nginx in front (1.15.8.1)

Edit: Using `websockets.unix_server`, with these parameters:  `close_timeout=.2, max_size=2**16, compression=None, backlog=65535`

**Comment by aaugustin** [comment]
I'm not quite sure what could explain this. 12k connections in a single process with 1k reconnections per minute sounds quite high for Python.

Regarding "things to check", I'd suggest instrumenting to check that `connection_lost` is called here:
 https://github.com/aaugustin/websockets/blob/8.1/src/websockets/protocol.py#L1349

Since you're iterating over incoming messages, your code is waiting for data here:
https://github.com/aaugustin/websockets/blob/8.1/src/websockets/protocol.py#L492-L496

The `self.transfer_data_task` coroutine is reading from the network and waiting for data here:
https://github.com/aaugustin/websockets/blob/8.1/src/websockets/protocol.py#L827

asyncio must call `connection_lost` when a TCP connection drops, which should be the case for connections "definitely disconnected from the system".

Through a not quite trivial but rather well tested sequence of events, this terminates `self.transfer_data_task`, and the `ConnectionClosed` exception is raised here:
https://github.com/aaugustin/websockets/blob/8.1/src/websockets/protocol.py#L509

If `connection_lost` was called, then the connection state is CLOSED, and `ensure_open` raises `ConnectionClosed` immediately.

**Comment by satchamo** [comment]
Thank you for the ideas. I'm going to add a little `debug_state` property to the websocket and update it throughout the close workflow in websockets and in my code. Hopefully that will pinpoint where things gets stuck. Unfortunately, this situation happens infrequently, so it could weeks before I see it again.

I tried to reproduce the problem by hammering the server with new connections (with 10k or 20k connections already open)...and it just chugs along fine! So I don't _think_ my problem is caused by the CPU not having enough time to get things done. Additionally, when these zombie connections start accumulating, the process seems to have no trouble accepting new connections (at least, for a while).

 But we shall see. I'll post back when I get some more information from production...

**Comment by aaugustin** [comment]
Well. This isn't going to be easy to debug :-/ Let me know if you find something!

**Comment by aaugustin** [comment]
Need more info before deciding what I can make of this.

**Comment by satchamo** [comment]
Well, I was barking up the wrong tree. I'm very sorry for wasting your time!

For posterity: I noticed a bunch of closed websockets stacking up again. I checked out the logs of my newly instrumented code, and all the websockets were stuck in my finally clause. So websockets **did** raise the exception. 

My websockets are stuck (somewhere) on these lines:
```python
with (await db_pool) as conn:
    cur = await conn.cursor()
```
where db_pool is a connection pool from the aiomysql library.

I never would have guessed that would be a problem. Perhaps I need more db connections in the pool. MySQL is perfectly capable of thousands of queries per second and I only do an update query when the connection is opened, and closed (and the query uses the primary index). At least I have a better idea now where the problem is.

Thank you again for the assistance, and thank you for all the work you put into this very elegant websocket API.
